### PR TITLE
Enhance status badge for PRs in [GithubIssueDetail]

### PR DESCRIPTION
**Pull request by ice1000**
I'm basically following other PRs -- I am not an expert on JS.
Please don't hesitate to spot any problems.

This PR addresses #1719

### CODE REVIEW

**Comment by ice1000** [comment]
There are some linter errors that I don't quite understand:
```
  53:27  error  Identifier 'merged_at' is not in camel case  camelcase
  53:38  error  Identifier 'closed_at' is not in camel case  camelcase
  55:9   error  Identifier 'merged_at' is not in camel case  camelcase
  57:9   error  Expected property shorthand                  object-shorthand
  61:16  error  Identifier 'closed_at' is not in camel case  camelcase
  63:9   error  Expected property shorthand                  object-shorthand
  69:9   error  Expected property shorthand                  object-shorthand
  77:13  error  Identifier 'merged_at' is not in camel case  camelcase
  77:24  error  Identifier 'closed_at' is not in camel case  camelcase
```

What if the API is using snake case?

**Comment by ice1000** [comment]
Trying to use this code:

```js
    return this.constructor.render({
      number,
      mergedAt: json['merged_at'],
      closedAt: json['closed_at'],
    })
```

instead of matching the properties out directly. Hope the CI will be happy with that.

**Comment by ice1000** [comment]
What I am thinking about is that the badge rendering logic is completely different and they're using different GitHub APIs.

Several questions:

+ Is it a good idea to have 2 API requests in the issue detail badge logic?
+ Isn't the badge rendering logic I've added are quite different from existing ones?

Yes, these questions are all present in the original issue, but my reason of adding a new badge is that I don't want to have 2 API requests in one badge and the logic is pretty different (because we're using different APIs!).

**Comment by calebcartwright** [comment]
> Is it a good idea to have 2 API requests in the issue detail badge logic?

No one is proposing making 2 sequential API calls. There's plenty of badge service implementations that use parameters to conditionally determine the URL call to make, the schema to use, how to transform the response, and how to render. 

Many such conditions are already handled within the existing service class.

**Comment by ice1000** [comment]
> route/query parameters to conditionally determine the URL call to make

This is very very reasonable! I've got a plan in mind, please comment:

+ Add a new `which` to issue service, like `pull`
+ If the `which` is `pull`, use the logic that I've implemented to determine the `message`

**Comment by calebcartwright** [comment]
> This is very very reasonable! I've got a plan in mind, please comment:
> 
> * Add a new `which` to the issue detail service, like `pull`
> * If the `which` is `pull`, use the logic that I've implemented to determine the `message`

That's probably the approach I'd suggest as well. Below are some snippets from what I was mocking up locally, may be a useful reference:

Update the state map:
https://github.com/badges/shields/blob/gh-pr-status/services/github/github-issue-detail.service.js#L21
```js
const stateMap = {
  schema: Joi.object({
    ...commonSchemaFields,
    state: Joi.string()
      .allow('open', 'closed')
      .required(),
    merged_at: Joi.string().allow(null),
  }).required(),
  transform: ({ json }) => ({
    state: json.state,
    merged: json.merged_at !== null,
  }),
  render: ({ value, isPR, number }) => {
    const state = value.state
    const label = `${isPR ? 'pull request' : 'issue'} ${number}`

    if (!isPR || state === 'open') {
      return {
        color: stateColor(state),
        label,
        message: state,
      }
    }

    if (value.merged) {
      return {
        label,
        message: 'merged',
        color: 'blueviolet',
      }
    }

    return {
      label,
      message: 'rejected',
      color: 'red',
    }
  },
}
```

the service class `route`:
```js
  static get route() {
    return {
      base: 'github',
      pattern:
        ':kind(issues|pulls)/detail/:which(state|title|author|label|comments|age|last-update)/:user/:repo/:number([0-9]+)',
    }
  }
```

the service class example will need to have the `namedParams` and `staticPreview` tweaked:
```js
...
        namedParams: {
          kind: 'issues',
          which: 'state',
          user: 'badges',
          repo: 'shields',
          number: '979',
        },
        staticPreview: this.render({
          which: 'state',
          value: { state: 'closed' },
          isPR: false,
          number: '979',
        }),
...
```

and the service implementations for `fetch`, `handle` and `transform` at the bottom of the file:
```js
  async fetch({ kind, which, user, repo, number }) {
    return this._requestJson({
      url: `/repos/${user}/${repo}/${kind}/${number}`,
      schema: whichMap[which].schema,
      errorMessages: errorMessagesFor('issue, pull request or repo not found'),
    })
  }

  transform({ json, which, kind }) {
    const value = whichMap[which].transform({ json, which })
    const isPR = json.hasOwnProperty('pull_request') || kind === 'pulls'
    return { value, isPR }
  }

  async handle({ kind, which, user, repo, number }) {
    const json = await this.fetch({ kind, which, user, repo, number })
    const { value, isPR } = this.transform({ json, which, kind })
    return this.constructor.render({ which, value, isPR, number })
  }
```

And lastly, the `redirector` that's currently remapping `pulls` to `issues` here: https://github.com/badges/shields/blob/gh-pr-status/services/github/github-issue-detail-redirect.service.js#L22 will need to be removed

**Comment by calebcartwright** [comment]
There's a few tests that would need to be updated as well:

* the [unit tests](https://github.com/badges/shields/blob/gh-pr-status/services/github/github-issue-detail.spec.js#L13) should have a couple more added to validate the new render logic and the [transform](https://github.com/badges/shields/blob/gh-pr-status/services/github/github-issue-detail.spec.js#L98) logic 
* in the [redirector tests](https://github.com/badges/shields/blob/gh-pr-status/services/github/github-issue-detail-redirect.tester.js) 
  * [this one](https://github.com/badges/shields/blob/gh-pr-status/services/github/github-issue-detail-redirect.tester.js#L11) can be deleted
  * [this one](https://github.com/badges/shields/blob/gh-pr-status/services/github/github-issue-detail-redirect.tester.js#L48) should be updated (change `issues` to `pulls`)
  * [this one](https://github.com/badges/shields/blob/gh-pr-status/services/github/github-issue-detail-redirect.tester.js#L58) should be updated (change `issues` to `pulls`)
* in the [live/service tests](https://github.com/badges/shields/blob/gh-pr-status/services/github/github-issue-detail.tester.js) we should add a test or two that use a real PR

**Comment by ice1000** [comment]
Done, please review.

**Comment by calebcartwright** [comment]
Excellent! I'll review here shortly

File: services/github/github-issue-detail-redirect.service.js [hunk 1]
```diff
@@ -27,7 +27,7 @@ module.exports = [
         ':which(state|title|author|label|comments|age|last-update)/:user/:repo/:number([0-9]+)',
     },
     transformPath: ({ which, user, repo, number }) =>
```
Comment (calebcartwright) [inline services/github/github-issue-detail-redirect.service.js]: Two items in this file:

1) this entire redirector (from lines 22-31) should be removed
2) the `trasformPath` response above on line 19 should be
` /github/${issueKind}/detail/${whichMap[which]}/${user}/${repo}/${number}`,

File: services/github/github-issue-detail.spec.js [hunk 2]
```diff
@@ -99,9 +99,26 @@ describe('GithubIssueDetail', function() {
       which: 'state',
       json: { state: 'closed' },
     }).expect({
-      value: 'closed',
+      value: { state: 'closed', merged: true }
```
Comment (calebcartwright) [inline services/github/github-issue-detail.spec.js]: the missing comma at the end of this line is the reason (or at least one of the reasons) why eslint, prettier and the unit tests are failing in the CI builds. 
```suggestion
      value: { state: 'closed', merged: true },
```

File: services/github/github-issue-detail.spec.js [hunk 3]
```diff
@@ -99,9 +99,26 @@ describe('GithubIssueDetail', function() {
...
       isPR: false,
     })
+    given({
+      which: 'state',
+      kind: 'pulls',
+      json: { state: 'closed' },
+    }).expect({
+      // Since it's a PR, the "merged" value is not crucial here.
+      value: { state: 'closed', merged: true }
```
Comment (calebcartwright) [inline services/github/github-issue-detail.spec.js]:
```suggestion
      value: { state: 'closed', merged: true },
```

**Comment by ice1000** [comment]
Addressed all comments, rebased on master, please take a look.

**Comment by calebcartwright** [comment]
Review app deployed https://shields-staging-pr-3295.herokuapp.com/

https://shields-staging-pr-3295.herokuapp.com/github/pulls/detail/state/badges/shields/3269.svg


Looking good!

**Pull request merged**